# **2. Tensors**

TensorFlow revolves around the concept of Tensors. A tensor is a multi-dimensional array, similar to a NumPy array, having different ranks (0D, 1D, 2D, etc..) representing scalars, vectors, matrices and higher-dimensional arrays.

**Properties of Tensors:**<br>
1. **Rank** - No. of dimensions
2. **Shape** - No. of elements along each dimension  e.g. 2x3 matrix has shape (2,3)
3. **Data Type** - Type of data  e.g. tf.float32, tf.int64, tf.string
4. **Immutable** - Values cannot be changed after creation

In [2]:
# Importing tensorflow
import tensorflow as tf

In [5]:
# Scalars (Rank 0)
scalar = tf.constant(3)
print("Scalar:", scalar)

# Vectors  (Rank 1)
vector = tf.constant([1,2,3])
print("Vector:", vector)

# Matrices (Rank 2)
matrix = tf.constant([[1,2,3],[4,5,6]])
print("Matrix:")
print(matrix)

# Higher dimensional Tensors (Rank > 2)
tensor_3d = tf.constant([[[1,2],[3,4]],[[5,6],[7,8]]])  # Assuming 3-D
print("3D Tensor:")
tensor_3d

Scalar: tf.Tensor(3, shape=(), dtype=int32)
Vector: tf.Tensor([1 2 3], shape=(3,), dtype=int32)
Matrix:
tf.Tensor(
[[1 2 3]
 [4 5 6]], shape=(2, 3), dtype=int32)
3D Tensor:


<tf.Tensor: shape=(2, 2, 2), dtype=int32, numpy=
array([[[1, 2],
        [3, 4]],

       [[5, 6],
        [7, 8]]])>

In [6]:
# Shape of Tensor
tensor_3d.shape

TensorShape([2, 2, 2])

In [7]:
# Data type of tensor
tensor_3d.dtype

tf.int32

In [8]:
# Original Tensor
print(tensor_3d)

tf.Tensor(
[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]], shape=(2, 2, 2), dtype=int32)


In [9]:
# Tensor after multiplying with 4
print(tensor_3d*4)

tf.Tensor(
[[[ 4  8]
  [12 16]]

 [[20 24]
  [28 32]]], shape=(2, 2, 2), dtype=int32)


In [10]:
# Changes not saved until equalized with tensor_3d again
print(tensor_3d)

tf.Tensor(
[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]], shape=(2, 2, 2), dtype=int32)


In [11]:
# Applying changes
tensor_3d = tensor_3d*4
tensor_3d

<tf.Tensor: shape=(2, 2, 2), dtype=int32, numpy=
array([[[ 4,  8],
        [12, 16]],

       [[20, 24],
        [28, 32]]])>

In [13]:
# Transpose of a matrix - Taking 2D for easy visualizing - applicable for all dimensions.
print(matrix)
tf.transpose(matrix)

tf.Tensor(
[[1 2 3]
 [4 5 6]], shape=(2, 3), dtype=int32)


<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[1, 4],
       [2, 5],
       [3, 6]])>

In [14]:
matrix

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[1, 2, 3],
       [4, 5, 6]])>

In [15]:
matrix+matrix

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[ 2,  4,  6],
       [ 8, 10, 12]])>

In [18]:
# Transpose of a matrix
tf.transpose(matrix)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[1, 4],
       [2, 5],
       [3, 6]])>

In [20]:
# Matrix multiplication - multiplying matrix with its transpose here.
matrix @ tf.transpose(matrix)    # (2,3) multiplied with (3,2) -> (2,2)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[14, 32],
       [32, 77]])>

In [21]:
# Concatenation of tensors

# Let's create sample tensors
tensor1 = tensor_3d
tensor2 = tensor_3d*5+2
tensor3 = tensor_3d*4+3

print(tensor1)
print(tensor2)
print(tensor3)

tf.Tensor(
[[[ 4  8]
  [12 16]]

 [[20 24]
  [28 32]]], shape=(2, 2, 2), dtype=int32)
tf.Tensor(
[[[ 22  42]
  [ 62  82]]

 [[102 122]
  [142 162]]], shape=(2, 2, 2), dtype=int32)
tf.Tensor(
[[[ 19  35]
  [ 51  67]]

 [[ 83  99]
  [115 131]]], shape=(2, 2, 2), dtype=int32)


In [52]:
# Concatenating
tensor_concat = tf.concat([tensor1, tensor2, tensor3], axis=0)
tensor_concat

<tf.Tensor: shape=(6, 2, 2), dtype=int32, numpy=
array([[[  4,   8],
        [ 12,  16]],

       [[ 20,  24],
        [ 28,  32]],

       [[ 22,  42],
        [ 62,  82]],

       [[102, 122],
        [142, 162]],

       [[ 19,  35],
        [ 51,  67]],

       [[ 83,  99],
        [115, 131]]])>

<br>The above snippet for concatenation brought contents of all the three 3-D tensors into one 3-D tensor. Look at the shape.

(2,2,2) -> A tensor having 2 elements of type 2x2 <br>
(6,2,2) -> A tensor having 6 elements of type 2x2<br>

In [34]:
# Applying Activation Function  -  Softmax

# 1. Initializing vector
vector1 = tf.constant([2., 1., .1])
print(vector1)

# 2. Applying softmax to logits
print(tf.nn.softmax(vector1))

tf.Tensor([2.  1.  0.1], shape=(3,), dtype=float32)
tf.Tensor([0.6590011  0.24243298 0.09856588], shape=(3,), dtype=float32)


<br> **Understanding Softmax Activation function:** <br>
Formula:  &emsp;e^(yi)/sum(e^(yj)) <br>

denominator = e^2 + e^1 + e^0.1 = 11.2125

softmax([2.0, 1.0, 0.1]) = [ans(2.0), ans(1.0), ans(0.1)]

ans(2.0) = e^(2.0)/denominator = (7.389)/11.2125 = 0.6590011
ans(1.0) = e^(1.0)/denominator = 0.2424
ans(0.1) = e^(0.1)/denominator = 0.0985

So, softmax([2.,1.,.1]) = [0.659011, 0.2424, 0.0985]

In [39]:
# Converting into NumPy array
tf.nn.softmax(vector1).numpy()

array([0.6590011 , 0.24243298, 0.09856588], dtype=float32)

In [40]:
# Converting into Python List
tf.nn.softmax(vector1).numpy().tolist()

[0.6590011119842529, 0.24243298172950745, 0.09856588393449783]

In [51]:
# Summation
tf.reduce_sum(vector1).numpy()

3.1

In [54]:
# Summation for higher-dimensions
tf.reduce_sum(tensor_concat).numpy()

1480

In [57]:
# Converting other iterables to tensor
tf.convert_to_tensor([1,2,3]) # More versatile than tf.constant([1,2,3])

<tf.Tensor: shape=(3,), dtype=int32, numpy=array([1, 2, 3])>

**What's difference between tf.constant() and tf.convert_to_tensor()?** <br>
1. tf.convert_to_tensor() is more versatile in terms of input types - Python List, NumPy arrays, and even a TensorFlow tensor. If input is a tensor already, it returns the input as it is. Whereas, tf.constant() is designed specifically to create tensors with constant fixed values.
2. tf.constant() is explicitly immutable. But the immutability of tf.convert_to_tensor() depends on the input type. If input is already a tensor and it was made of tf.constant() it will be returned as it is, making it immutable.

<br><br> **GPU Configuration for faster computation**

While performing complex deep learning tasks, CPUs may load longer to complete the tasks. In order to speed up the tasks, one may use GPUs if it is available with them. It is said that combining GPUs with high-end computer components can render graphics up to 100 times faster than CPUs. If you have one, here's the way to configure

In [69]:
# Checking availability of GPU
if tf.config.list_physical_devices("GPU"):
    print("GPU is present")
else:
    print("GPU is NOT present")

GPU is NOT present


In [71]:
# List available physical devices (GPUs)
devices = tf.config.list_physical_devices("GPU")

if devices:
    # Set memory growth for each GPU (if available)
    for device in devices:
        tf.config.experimental.set_memory_growth(device, True)
        print(f"Memory growth set for {device}")
else:
    print("No GPU devices found.")

No GPU devices found.
